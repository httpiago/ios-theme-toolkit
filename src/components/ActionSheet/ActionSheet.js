import React from 'react'
import PropTypes from 'prop-types'
import { mapClassList } from '../../Utils'
import Drawer from '../Drawer/Drawer'
import ButtonGroup from '../ButtonGroup/ButtonGroup'

/**
 * ActionSheet component.
 *
 * @since 0.5.0
 * @see https://httpiago.github.io/ios-theme-toolkit/#/components/ActionSheet
 */
export default function ActionSheet({ content, buttons, className: aditionalClasses, closeOnInnerClick, onRequestClose, ...rest }) {

  const prefix = 'action-sheet'
  //const classes = mapClassList({})

  const handleInnerClick = () => {
    if (closeOnInnerClick === true) onRequestClose()
  }

  return (
    <Drawer
      className={`${prefix} ${aditionalClasses}`}
      {...rest}
      placement="bottom"
      backdrop
      closeByBackdropClick={true}
      onRequestClose={onRequestClose}
      onClick={handleInnerClick}
      content={(
        <React.Fragment>
          {content && <div className="action-sheet__content">{content}</div>}

          {buttons && (
            <ButtonGroup
              vertical={true}
              className="modal__buttons"
            >
              {React.Children.toArray(buttons)}
            </ButtonGroup>
          )}
        </React.Fragment>
      )}
    />
  )
}

ActionSheet.defaultProps = {
  className: '',
  content: null,
  closeOnInnerClick: true
}

ActionSheet.propTypes = {
  /** Aditional classes. */
  className: PropTypes.string,
  /** Content above the buttons */
  content: PropTypes.node,
  /** Buttons */
  buttons: PropTypes.arrayOf(PropTypes.node),
  /** Function that will be called when the user clicks outside of elment and in the future when pressing the esc key */
  onRequestClose: PropTypes.func.isRequired,
  /** Close when user click inside action sheet */
  closeOnInnerClick: PropTypes.bool,
}
