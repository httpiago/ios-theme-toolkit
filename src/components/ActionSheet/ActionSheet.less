@import '../../customizations.less';

// Default style
.drawer.action-sheet {
  background-color: #fafafa;
  border-radius: 13px;
  box-shadow: none;
  padding: 0;
  margin: -@base-margin auto 0;
  max-width: 460px;
  left: @base-margin;
  right: @base-margin;

  &:not(.drawer--visible) {
    margin-top: 0;
  }
}

.action-sheet__content {
  font-size: @font-size;
  text-align: center;
  padding: @base-padding;
  color: #8a8a8a;
}
