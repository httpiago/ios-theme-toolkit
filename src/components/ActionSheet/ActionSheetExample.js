import React from 'react'
import ActionSheet from './ActionSheet'
import Button from '../Button/Button'

export default class ASE extends React.Component {
  state = {
    visibility1: false,
    visibility2: false,
    visibility3: false,
  }

  toggleActionsheet = (id, val) => {
    let stateName = `visibility${id}`
    this.setState({ [stateName]: val })
  }

  render() {
    return (
      <React.Fragment>
        <ActionSheet
          visible={this.state.visibility1}
          buttons={[
            <Button>Add to playlist</Button>,
            <Button>Copy link</Button>,
            <Button>Archive</Button>,
            <Button color="red">Delete</Button>
          ]}
          onRequestClose={() => this.toggleActionsheet(1, false)}
        />

        <Button fill onClick={() => this.toggleActionsheet(1, true)}>Open ActionSheet</Button>

        <br/><br/>

        <ActionSheet
          visible={this.state.visibility2}
          content={'Amet tempor nostrud dolore non consequat elit do proident irure sit irure exercitation.'}
          onRequestClose={() => this.toggleActionsheet(2, false)}
        />

        <Button fill onClick={() => this.toggleActionsheet(2, true)}>Open ActionSheet without buttons</Button>

        <br/><br/>

        <ActionSheet
          visible={this.state.visibility3}
          content={'What do you want to do?'}
          buttons={[
            <Button>Add to playlist</Button>,
            <Button>Copy link</Button>,
            <Button>Archive</Button>,
            <Button color="red">Delete</Button>
          ]}
          onRequestClose={() => this.toggleActionsheet(3, false)}
        />

        <Button fill onClick={() => this.toggleActionsheet(3, true)}>Open ActionSheet with title</Button>
      </React.Fragment>
    )
  }
}
