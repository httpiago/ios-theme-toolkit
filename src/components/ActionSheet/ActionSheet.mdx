---
name: ActionSheet
route: /components/ActionSheet
menu: Components
---

import { Playground, PropsTable } from 'docz'
import { ActionSheet } from '../../index.js'
import '../../../dist/styles.css'
import ActionSheetExample from './ActionSheetExample'

# ActionSheet

To use this component, simply import it into any file and pass the [modifiers](#/#props) in the properties.

Note: This component has no state (a controlled component), you must provide the visibility state of the component and close it "manually".

See more in the example below:

```jsx
import { ActionSheet, Button } from 'ios-theme-toolkit'

import { Drawer, Button } from 'ios-theme-toolkit'

export default class ActionSheetExample extends React.Component {
  state = {
    visibility: false,
  }

  openActionSheet = () => {
    this.setState({ visibility: true })
  }

  closeActionSheet = () => {
    this.setState({ visibility: false })
  }

  render() {
    return (
      <React.Fragment>
        <ActionSheet
          visible={this.state.visibility}
          buttons={[
            <Button>Add to</Button>,
            <Button>Copy link</Button>,
            <Button color="red">Delete</Button>
          ]}
          onRequestClose={this.closeActionSheet}
        />

        <Button fill onClick={this.openActionSheet}>Open Action Sheet</Button>
      </React.Fragment>
    )
  }
}
```

## Examples

<Playground>
  <ActionSheetExample />
</Playground>

## Props

<PropsTable of={ ActionSheet } />

All the properties which are not listed above will be transferred to the [Drawer component](#/components/Drawer).
