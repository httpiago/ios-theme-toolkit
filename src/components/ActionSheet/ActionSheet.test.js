/**
 * TEST FILE
 * Please try writing the tests for all component behaviors,
 * manually call the functions within the class, change the properties...
 *
 * @see https://jestjs.io/docs/en/expect
 * @see https://github.com/airbnb/enzyme/tree/master/docs/api
 */
import React from 'react'
import { shallow } from 'enzyme'
import ActionSheet from './ActionSheet'

const COMPONENT = shallow(
  <ActionSheet
    className="aditionalClass"
    visible={true}
    onRequestClose={() => {}}
  />
)

describe('Test ActionSheet component', () => {
  it('should render correctly based on last snapshot', () => {
    expect(COMPONENT).toMatchSnapshot()
  })

  // it('should update correctly', () => {
  //   COMPONENT.setProps({ color: "green" })

  //   expect(COMPONENT.hasClass('button--green')).toBeTruthy()
  // })

  // it('should renders children correctly', () => {
  //   expect(COMPONENT.text()).toContain('content test')
  // })

  // it('should has aditional classes', () => {
  //   expect(COMPONENT.hasClass('aditionalClass')).toBeTruthy()
  // })

  // it('should pass aditional props to html element', () => {
  //   COMPONENT.setProps({ 'data-custom-attr': 'yes' })

  //   expect(COMPONENT.html()).toContain('data-custom-attr="yes"')
  // })

})
