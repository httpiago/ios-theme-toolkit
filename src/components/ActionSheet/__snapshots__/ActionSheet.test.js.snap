// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ActionSheet component should render correctly based on last snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ActionSheet
    buttons={
      Array [
        <div
          className="button"
        />,
      ]
    }
    className="aditionalClass"
    closeOnInnerClick={true}
    content={null}
    onRequestClose={[Function]}
    visible={true}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "backdrop": true,
      "className": "action-sheet aditionalClass",
      "closeByBackdropClick": true,
      "content": <React.Fragment>
        <ButtonGroup
          className="modal__buttons"
          outline={false}
          round={false}
          vertical={true}
        >
          <div
            className="button"
          />
        </ButtonGroup>
      </React.Fragment>,
      "onClick": [Function],
      "onRequestClose": [Function],
      "placement": "bottom",
      "visible": true,
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "backdrop": true,
        "className": "action-sheet aditionalClass",
        "closeByBackdropClick": true,
        "content": <React.Fragment>
          <ButtonGroup
            className="modal__buttons"
            outline={false}
            round={false}
            vertical={true}
          >
            <div
              className="button"
            />
          </ButtonGroup>
        </React.Fragment>,
        "onClick": [Function],
        "onRequestClose": [Function],
        "placement": "bottom",
        "visible": true,
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
